### ISSUE TITLE
gcc14 condition coverage exception "The number of conditions must be equal."

### ISSUE THREAD

**Issue by dlegaultbbry** [gcovr/gcovr#1005]
# Describe the bug

I see this currently with the new rollout of gcovr 8.0 and gcc14 with the new conditions coverage in our environment. Still trying to narrow down why it's happening and what code is causing it.

```
gcovr --root=. --decisions -e '/.*/unittests/' --txt-summary --json --output unittest.json
(INFO) Reading coverage data...
Traceback (most recent call last):
  File "/home/vulcan/.local/lib/python3.10/site-packages/gcovr/formats/gcov/workers.py", line 81, in worker
    work(*args, **kwargs)
  File "/home/vulcan/.local/lib/python3.10/site-packages/gcovr/formats/gcov/read.py", line 566, in process_datafile
    done = run_gcov_and_process_files(
  File "/home/vulcan/.local/lib/python3.10/site-packages/gcovr/formats/gcov/read.py", line 922, in run_gcov_and_process_files
    process_gcov_json_data(gcov_filename, covdata, options)
  File "/home/vulcan/.local/lib/python3.10/site-packages/gcovr/formats/gcov/read.py", line 301, in process_gcov_json_data
    insert_file_coverage(covdata, file_cov, get_merge_mode_from_options(options))
  File "/home/vulcan/.local/lib/python3.10/site-packages/gcovr/merging.py", line 202, in insert_file_coverage
    return _insert_coverage_item(target, file.filename, file, merge_file, options)
  File "/home/vulcan/.local/lib/python3.10/site-packages/gcovr/merging.py", line 180, in _insert_coverage_item
    merged_item = merge_item(target_dict[key], new_item, options)
  File "/home/vulcan/.local/lib/python3.10/site-packages/gcovr/merging.py", line 221, in merge_file
    left.lines = _merge_dict(left.lines, right.lines, merge_line, options)
  File "/home/vulcan/.local/lib/python3.10/site-packages/gcovr/merging.py", line 144, in _merge_dict
    _insert_coverage_item(left, key, right_item, merge_item, options)
  File "/home/vulcan/.local/lib/python3.10/site-packages/gcovr/merging.py", line 180, in _insert_coverage_item
    merged_item = merge_item(target_dict[key], new_item, options)
  File "/home/vulcan/.local/lib/python3.10/site-packages/gcovr/merging.py", line 262, in merge_line
    left.conditions = _merge_dict(
  File "/home/vulcan/.local/lib/python3.10/site-packages/gcovr/merging.py", line 144, in _merge_dict
    _insert_coverage_item(left, key, right_item, merge_item, options)
  File "/home/vulcan/.local/lib/python3.10/site-packages/gcovr/merging.py", line 180, in _insert_coverage_item
    merged_item = merge_item(target_dict[key], new_item, options)
  File "/home/vulcan/.local/lib/python3.10/site-packages/gcovr/merging.py", line 471, in merge_condition
    raise AssertionError("The number of conditions must be equal.")
AssertionError: The number of conditions must be equal.
```

# To Reproduce

Unknown yet, gathering data.

# Expected behaviour

No exception.

# Desktop (please complete the following information)

Ubuntu 22.04 with gcc14.2

# Additional context

**Comment by Spacetown** [comment]
It seems that you get for the same line different count of conditions.
Do you compile the same file with a different set of preprocessing definitions which affect the conditions? Do you have a function within a condition in a header one time unlined and one time not?

**Comment by Spacetown** [comment]
> Is one compile unit having condition coverage enabled and the other haven't?
This doesn't trigger the issue. Here the file has is missing only the condition stats.

**Comment by dlegaultbbry** [comment]
All the files are compiled with the same options with an opt level of -O0 since this is unittesting and they all have condition coverage enabled as it detects if the compiler supports it or not.

I'll provide more details when I manage to extract which file and code is problematic.

The way we do this is that each function is tested separately like this:

Testing func1 in `project/path/to/real/file.c`
Associated test `project/unittests/path/to/real/file.c/func1_test.c`
```
// override function in source to control its behaviour
static void my_stub_func(...) { ... }
#define stub_func my_stub_func

#include source.c <-- this contains an #include "path/to/real/file.c"

int main()
{
  // call func1 with all combinations of parameters for coverage returning 0 or 1 on PASS or FAIL
}
```

**Comment by Spacetown** [comment]
I think we need to improve the logging for such errors. I'll play a little bit with tracking additional infos in verbose mode.

**Comment by dlegaultbbry** [comment]
Sounds good. if you have a branch for me to try and get more information quickly that would be great. I'm a bit swamped in fires for the rest of the week but will pick this up next Monday to get to the bottom of what is happening.

**Comment by Spacetown** [comment]
Please use https://github.com/Spacetown/gcovr/tree/improve_logging. If it helps I'll create a PR for it.
The source data source files are added to the internal data model printed on an `AssertionError` while merging the `FileCoverage`. In verbose mode the stat is also added to the JSON trace file. If a trace file is read in and the keys is missing the name of the trace file is used as data source.

**Comment by dlegaultbbry** [comment]
The line of code in question is of the form below. All other coverage lines around it are always reported with the right information for this function.

```
for (A = 0; A < NUM; ++A) { 
   if ((A != B) && (C != D)) { <-- problem line
     ....
   }
}
```

For all tests except one, it properly reports a count of 4 which is accurate, but in one test it sees it as 6.

```
-lineno: 397 - condition {'count': 6, 'covered': 0, 'not_covered_true': [0, 1, 2], 'not_covered_false': [0, 1, 2]}
+lineno: 397 - condition {'count': 4, 'covered': 0, 'not_covered_true': [0, 1], 'not_covered_false': [0, 1]}

```

One thing that would be really convenient in debug is printing the actual line of code where the problem occurred in merging. I had to hack things away some more to dump all the lines with conditions and then diff between one test showing 4 and the other showing 6 to find the offending line.

I've fixed it locally by adding a temporary merge option to just squeeze the size of the elements so they are equal (whichever is larger gets reduced to the size of the other). Not sure what else can be done than adding another ignore this type of error message with a warning for now. With this, the merge works and the reported data still looks accurate.

I've looked into gcov bug history but can't see anything related to conditions yet but that may just be because not a lot of people use this right now.

**Comment by Spacetown** [comment]
Please can you check the last change on my branch with the improved logging? When merging line data the line information is added to the assertion.

**Comment by dlegaultbbry** [comment]
I tried the changes and they work fine.

**Comment by Spacetown** [comment]
Fixed with #1009.
